core/src/main/java/org/hisp/dhis/android/core/trackedentity/internal/TrackedEntityInstanceDownloadCall.kt

```
@@ -133,19 +138,68 @@ internal class TrackedEntityInstanceDownloadCall(
         useEntityEndpoint: Boolean,
         query: TrackerAPIQuery,
     ): Result<TrackedEntityInstance?, D2Error> {
-        return if (useEntityEndpoint) {
-            coroutineCallExecutor.wrap(
-                storeError = true,
-                errorCatcher = TrackedEntityInstanceCallErrorCatcher(),
-            ) {
-                trackerCallFactory.getTrackedEntityCall().getEntityCall(uid, query)
-            }
-        } else {
+        if (!useEntityEndpoint) {
             val collectionQuery = query.copy(uids = listOf(uid))
-            coroutineCallExecutor.wrap(storeError = true) {
+            return coroutineCallExecutor.wrap(storeError = true) {
                 trackerCallFactory.getTrackedEntityCall().getCollectionCall(collectionQuery)
             }.map { it.items.firstOrNull() }
         }
+
+        val result = coroutineCallExecutor.wrap(
+            storeError = true,
+            errorCatcher = TrackedEntityInstanceCallErrorCatcher(),
+        ) {
+            trackerCallFactory.getTrackedEntityCall().getEntityCall(uid, query)
+        }
+
+        return when {
+            result is Result.Success -> result
+            result is Result.Failure &&
+                result.failure.errorCode() == D2ErrorCode.OWNERSHIP_ACCESS_DENIED -> result
+            result is Result.Failure &&
+                result.failure.errorCode() == D2ErrorCode.PROGRAM_ACCESS_CLOSED -> result
+            result is Result.Failure && result.failure.httpErrorCode() == HttpStatusCode.NotFound.value -> {
+                checkOwnershipOnNotFound(uid, query)
+            }
+            else -> result
+        }
+    }
+
+    private suspend fun checkOwnershipOnNotFound(
+        uid: String,
+        query: TrackerAPIQuery,
+    ): Result<TrackedEntityInstance?, D2Error> {
+        val program = query.commonParams.program ?: return Result.Success(null)
```

I think that it should return the original error if the program is not defined (instead of null).